File: src/main/java/com/partymakers/shareparty/domain/party/usecase/impl/RemovePartyExpenseImpl.java
package com.partymakers.shareparty.domain.party.usecase.impl;

import com.partymakers.shareparty.domain.party.entity.Expense;
import com.partymakers.shareparty.domain.party.entity.PartyRoom;
import com.partymakers.shareparty.domain.party.port.PartyRoomRepository;
import com.partymakers.shareparty.domain.party.usecase.RemovePartyExpense;
import com.partymakers.shareparty.domain.party.usecase.exception.NotFoundException;

import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
public class RemovePartyExpenseImpl implements RemovePartyExpense {

    private final PartyRoomRepository partyRoomRepository;

    @Override
    public void removePartyExpense(Long partyId, Expense expense) {
        partyRoomRepository.findById(partyId).ifPresentOrElse(
            partyRoom -> {
                partyRoom.getExpenses().removeIf(partyExpense -> partyExpense.equals(expense));
                partyRoomRepository.save(partyRoom);
            },
            () -> {throw new NotFoundException();});
    }
}

File: src/main/kotlin/com/partymakers/shareparty/domain/party/usecase/impl/RemovePartyExpenseImpl.kt
package com.partymakers.shareparty.domain.party.usecase.impl

import com.partymakers.shareparty.domain.party.entity.Expense
import com.partymakers.shareparty.domain.party.port.PartyRoomRepository
import com.partymakers.shareparty.domain.party.usecase.RemovePartyExpense
import com.partymakers.shareparty.domain.party.usecase.exception.NotFoundException

class RemovePartyExpenseImpl(
    private val partyRoomRepository: PartyRoomRepository
) : RemovePartyExpense {

    override fun removePartyExpense(partyId: Long, expense: Expense) {
        val partyRoom = partyRoomRepository.findById(partyId)
            .orElseThrow { NotFoundException("Party room with id: $partyId not found") }

        val updatedPartyRoom = partyRoom.copy(
            expenses = partyRoom.expenses.filterNot { it == expense }
        )
        partyRoomRepository.save(updatedPartyRoom)
    }
} 